Emby.Server.Implementations/Session/SessionWebSocketListener.cs

```diff
@@ -99,5 +130,130 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
             var controller = (WebSocketController)controllerInfo.Item1;
             controller.AddWebSocket(connection);
         }
+
+        /// <summary>
+        /// Called when a WebSocket is closed.
+        /// </summary>
+        /// <param name="sender">The WebSocket.</param>
+        /// <param name="e">The event arguments.</param>
+        private void _webSocket_Closed(object sender, EventArgs e)
+        {
+            var webSocket = (IWebSocketConnection) sender;
+            webSocket.Closed -= _webSocket_Closed;
+            _webSockets.TryRemove(webSocket, out _);
+        }
+
+        /// <summary>
+        /// Adds a WebSocket to the KeepAlive watchlist.
+        /// </summary>
+        /// <param name="webSocket">The WebSocket to monitor.</param>
+        private async void KeepAliveWebSocket(IWebSocketConnection webSocket)
+        {
+            _webSockets.TryAdd(webSocket, 0);
+            webSocket.Closed += _webSocket_Closed;
+            webSocket.LastKeepAliveDate = DateTime.UtcNow;
+
+            // Notify WebSocket about timeout
+            try
+            {
+                await SendForceKeepAlive(webSocket);
+            }
+            catch (WebSocketException exception)
+            {
+                _logger.LogDebug(exception, "Error sending ForceKeepAlive message to WebSocket.");
```

Comment:
I see an error as something that originates from inside the server itself.
In this case, I would consider a network failure not that relevant to the server as a whole. I mean, it's not rare for a connection to drop (be it for network issues or because the browser has been closed), thereby the server should handle it gracefully but not complain that much.

Also, `ForceKeepAlive` is sent when the client hasn't responded in a while, so the server is already expecting a possible failure.

I would have opted for a warning if it wasn't for the fact that the dropped connection will be notified to the interested session that could potentially become offline. So the idea was to avoid two messages caused by the same event.

Don't know if this makes sense :sweat_smile:. These were my considerations, but if you think that being more verbose is better then I would agree as well.

Reply:
Nevermind, I've mistaken that line for the one that comes after. In this case it does make sense to print an error as the connection has just been established.